docs/topics/whatsnew21.md

```diff
@@ -10,10 +23,1562 @@ All you need to do is [change the Kotlin version](configure-build-for-eap.md) to
+A user can opt in like this:
+
+```kotlin
+@OptIn(UnstableApi::class)
+interface MyImplementation: CoreLibraryApi
+```
+
+> When you use the `@SubclassOptInRequired` annotation to require opt-in, 
+> the requirement is not propagated to any [inner or nested classes](nested-classes.md).
+>
+{type="note"}
+
+For a real-world example of how to use the `@SubclassOptInRequired` annotation in your API,
+check out the [`SharedFlow`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-shared-flow/)
+interface in the `kotlinx.coroutines` library.
+
+### Improved overload resolution for functions with generic types
+
+Previously, if you had a number of overloads for a function where some had value parameters of a generic type
+and others had function types in the same position, the resolution behavior could sometimes be inconsistent.
+
+This led to different behavior depending on whether your overloads were member functions or extension functions.
+For example:
+
+```kotlin
+class KeyValueStore<K, V> {
+    fun store(key: K, value: V) {} // 1
+    fun store(key: K, lazyValue: () -> V) {} // 2
+}
+
+fun <K, V> KeyValueStore<K, V>.storeExtension(key: K, value: V) {} // 1 
+fun <K, V> KeyValueStore<K, V>.storeExtension(key: K, lazyValue: () -> V) {} // 2
+
+fun test(kvs: KeyValueStore<String, Int>) {
+    // Member functions
+    kvs.store("", 1)    // Resolves to 1
+    kvs.store("") { 1 } // Resolves to 2
+
+    // Extension functions
+    kvs.storeExtension("", 1)    // Resolves to 1
+    kvs.storeExtension("") { 1 } // Doesn't resolve
+}
+```
+
+In this example, the `KeyValueStore` class has two overloads for the `store()` function,
+where one overload has function parameters with generic types `K` and `V`,
+and another has a lambda function that returns a generic type `V`.
+Similarly, there are two overloads for the extension function: `storeExtension()`.
+
+When the `store()` function was called with and without a lambda function,
+the compiler successfully resolved the correct overloads.
+However, when the extension function `storeExtension()` was called with a lambda function,
+the compiler didn't resolve the correct overload because it incorrectly considered both overloads to be applicable.
+
+To fix this problem, we've introduced a new heuristic so that the compiler can discard a possible overload
+when a function parameter with a generic type can't accept a lambda function based on information from a different argument.
+This change makes the behavior of member functions and extension functions consistent,
+and it is enabled by default in Kotlin 2.1.0.
+
+### Improved exhaustiveness checks for when expressions with sealed classes
+
+In previous versions of Kotlin, the compiler required an `else` branch in `when`
+expressions for type parameters with sealed upper bounds, even when all cases in the `sealed class` hierarchy were covered.
+This behavior is addressed and improved in Kotlin 2.1.0,
```

Comment:
```suggestion

This behavior is addressed and improved in Kotlin 2.1.0,
```

Reply:
Well, not sure about this change.
Could you provide the motivation behind it? All other cases like this one remain the same...